internal/checks/latency.go

```diff
@@ -0,0 +1,484 @@
+}
+
+func (c *LatencyCheck) GetUnhealthyReason(methods []string) conf.UnhealthyReason {
+	if c.errorCircuitBreaker.IsOpen() {
+		c.logger.Debug(
+			"LatencyCheck is not passing due to too many errors.",
+			zap.String("upstreamID", c.upstreamConfig.ID),
+			zap.Error(c.Err),
+		)
+
+		return conf.ReasonErrorRate
+	}
+
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	// Only consider the passed methods, even if other methods' circuit breakers might be open.
+	//
+	// TODO(polsar): If the circuit breaker for any of the passed methods is open, we consider this upstream
+	//  as unhealthy for all the other passed methods, even if their circuit breakers are closed. This might
+	//  be undesirable, though since all the methods are part of the same request, we would have to somehow
+	//  modify the request to only contain the healthy methods. This seems like more trouble than is worth.
+	for _, method := range methods {
+		if breaker, exists := c.methodLatencyBreaker[method]; exists && breaker.IsOpen() {
+			c.logger.Debug(
+				"LatencyCheck is not passing due to high latency of an RPC method.",
+				zap.String("upstreamID", c.upstreamConfig.ID),
+				zap.Any("method", method),
+				zap.Error(c.Err),
+			)
+
+			return conf.ReasonLatencyTooHighRate
+		}
+	}
+
+	return conf.ReasonUnknownOrHealthy
+}
+
+func (c *LatencyCheck) RecordRequest(data *types.RequestData) {
+	if c.routingConfig.PassiveLatencyChecking {
+		return
+	}
+
+	latencyCircuitBreaker := c.getLatencyCircuitBreaker(data.Method)
+	latencyCircuitBreaker.RecordLatency(data.Latency)
+
+	if data.Latency >= latencyCircuitBreaker.GetThreshold() {
+		c.metricsContainer.LatencyCheckHighLatencies.WithLabelValues(
+			c.upstreamConfig.ID,
+			c.upstreamConfig.HTTPURL,
+			metrics.HTTPRequest,
+		).Inc()
+	}
+
+	if data.HTTPResponseCode >= http.StatusBadRequest {
+		// No RPC responses are available since the HTTP request errored out.
+		// TODO(polsar): We might want to emit a Prometheus stat like we do for an RPC error below.
+		c.errorCircuitBreaker.RecordRequest(c.isError(
+			strconv.Itoa(data.HTTPResponseCode),
+			"",
+			"",
+		)) // HTTP request error
+	} else if data.ResponseBody != nil {
+		for _, resp := range data.ResponseBody.GetSubResponses() {
+			if resp.Error != nil {
+				// Do not ignore this response even if it does not correspond to an RPC request.
+				if c.isError("", strconv.Itoa(resp.Error.Code), resp.Error.Message) {
+					c.metricsContainer.LatencyCheckErrors.WithLabelValues(
+						c.upstreamConfig.ID,
+						c.upstreamConfig.HTTPURL,
+						metrics.HTTPRequest,
+					).Inc()
+
+					// Even though this is a single HTTP request, we count each RPC JSON subresponse error.
+					c.errorCircuitBreaker.RecordRequest(true) // JSON RPC subrequest error
+				} else {
+					c.errorCircuitBreaker.RecordRequest(false) // JSON RPC subrequest OK
+				}
+			}
+		}
+
+		c.errorCircuitBreaker.RecordRequest(false) // HTTP request OK
```

Comment:
Is this an extra call to `RecordRequest`? All responses should be recorded in within the for loop over `GetSubResponses`